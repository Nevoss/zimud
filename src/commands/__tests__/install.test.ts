import path from 'path';
import { vol } from 'memfs';
import { getLog, runCommand } from 'test-utils';
import { exec } from 'child_process';

jest.mock('child_process', () => ({
	exec: jest.fn((_, callback) => callback()),
}));

describe('zimud install', () => {
	it('should fail when there is no config file', async () => {
		// Act.
		await runCommand('zimud install');

		// Assert.
		expect(getLog()).toMatchSnapshot();
		expect(exec).not.toHaveBeenCalled();
	});

	it('should fail when config file is invalid', async () => {
		// Arrange.
		vol.writeFileSync('.zimud.json', '{"invalid": "schema"}');

		// Act & Assert.
		await expect(runCommand('zimud install')).rejects.toThrowError(
			'config file is invalid'
		);

		expect(exec).not.toHaveBeenCalled();
	});

	it('should fail when the pacakge is disabled', async () => {
		// Arrange.
		vol.writeFileSync('.zimud.json', '{"enabled": false, "packages": []}');

		// Act.
		await runCommand('zimud install');

		// Assert.
		expect(getLog()).toMatchSnapshot();
		expect(exec).not.toHaveBeenCalled();
	});

	it('should warn when there are invalid packages', async () => {
		// Arrange.
		vol.mkdirSync('invalid-package');

		vol.writeFileSync('invalid-package/package.json', 'not-a-json');

		vol.writeFileSync(
			'.zimud.json',
			'{"enabled": true, "packages": ["invalid-package"]}'
		);

		// Act.
		await runCommand('zimud install');

		// Assert.
		const normalizedLog = getLog()
			.replace(/\\/g, '/') // Windows path delimiters.
			.replace(/[A-Z]:/g, ''); // Windows drive letter.

		expect(normalizedLog).toMatchSnapshot();
		expect(exec).not.toHaveBeenCalled();
	});

	it('should not install the packages when passing `--dry-run`', async () => {
		// Arrange.
		vol.mkdirSync('valid-package');

		vol.writeFileSync(
			'valid-package/package.json',
			'{"name": "valid-package"}'
		);

		vol.writeFileSync(
			'.zimud.json',
			'{"enabled": true, "packages": ["valid-package"]}'
		);

		// Act.
		await runCommand('zimud install --dry-run');

		// Assert.
		expect(getLog()).toMatchSnapshot();
		expect(exec).not.toHaveBeenCalled();
	});

	it('should install the packages', async () => {
		// Arrange.
		const packagesDir = 'packages';
		const packages = ['package-a', 'package-b'];

		packages.forEach((packageName) => {
			vol.mkdirSync(path.resolve(packagesDir, packageName), {
				recursive: true,
			});

			vol.writeFileSync(
				path.resolve(packagesDir, `${packageName}/package.json`),
				`{"name": "${packageName}"}`
			);
		});

		vol.writeFileSync(
			'.zimud.json',
			'{"enabled": true, "packages": ["packages/*"]}'
		);

		// Mock hrtime.
		const originalHrtime = process.hrtime;

		(process.hrtime as () => [number, number]) = jest.fn(() => [0, 0]);

		// Act.
		await runCommand('zimud install');

		// Assert.
		expect(getLog()).toMatchSnapshot();

		const expectedPackages = packages
			.reverse() // glob reverses the order.
			.map((packageName) => path.resolve(packagesDir, packageName))
			.join(' ');

		expect(exec).toHaveBeenCalledTimes(1);

		expect(exec).toHaveBeenCalledWith(
			`npm install ${expectedPackages} --no-package-lock --no-save`,
			expect.any(Function)
		);

		// Cleanup.
		process.hrtime = originalHrtime;
	});
});
