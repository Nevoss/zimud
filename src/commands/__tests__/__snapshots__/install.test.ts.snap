// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`zimud install should fail when the pacakge is disabled 1`] = `
"
zimud is disabled. change the 'enable' key to 'true' in '.zimud.json' file if you wish.
"
`;

exports[`zimud install should fail when there is no config file 1`] = `
"
zimud config file was not found. Run 'zimud init' if you wish to generate a new one.
"
`;

exports[`zimud install should install the packages 1`] = `
"
zimud has successfully installed the following packages (in 0s 0ms):
   - package-b
   - package-a
"
`;

exports[`zimud install should not install the packages when passing \`--dry-run\` 1`] = `
"
The following packages would have been installed:
   - valid-package
"
`;

exports[`zimud install should warn when there are invalid packages 1`] = `
"
No valid packages to install.

Some packages are not installed because they are invalid. Please check the following packages:
   - /invalid-package
"
`;
